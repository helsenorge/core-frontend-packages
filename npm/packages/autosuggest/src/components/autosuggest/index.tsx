import * as React from 'react';

import ReactAutosuggest, { AutosuggestPropsBase, InputProps, RenderSuggestion } from 'react-autosuggest';

import ErrorWrapper from '@helsenorge/designsystem-react/components/ErrorWrapper';
import { renderLabel } from '@helsenorge/designsystem-react/components/Label';
import { FormMode } from '@helsenorge/designsystem-react/constants';
import { useUuid } from '@helsenorge/designsystem-react/hooks/useUuid';
import { getAriaDescribedBy } from '@helsenorge/designsystem-react/utils/accessibility';

import styles from './styles.module.scss';

export interface Suggestion {
  label: string;
  value: string;
  optionalLabel?: string;
}

interface Props<T>
  extends Pick<
    AutosuggestPropsBase<T>,
    'focusInputOnSuggestionClick' | 'onSuggestionsClearRequested' | 'onSuggestionSelected' | 'onSuggestionsFetchRequested'
  > {
  /** En array med suggestions som skal vises */
  suggestions: Array<T>;
  /** Props som sendes ned til ReactAutosuggest */
  inputProps: InputProps<T>;
  /** ClassName som skal settes på wrapper */
  className?: string;
  /** Label som vises over ReactAutosuggest komponentet */
  label?: React.ReactNode;
  /** Activates Error style for the input */
  error?: boolean;
  /** Error text to show above the component */
  errorText?: string;
  /** Error text to show above the component */
  errorTextId?: string;
  /** Class name to pass to ErrorWrapper */
  errorWrapperClassName?: string;
  /** Tilpasset funksjon for å definere markup'en som rendres for en suggestion i suggestionslisten */
  renderSuggestion?: RenderSuggestion<T>;
}

const renderSuggestion = (suggestion: Suggestion): React.JSX.Element => (
  <span>
    {suggestion.label}
    {suggestion.optionalLabel && (
      <>
        {' '}
        <em>{suggestion.optionalLabel}</em>
      </>
    )}
  </span>
);

const getSuggestionValue = (suggestion: Suggestion): string => (suggestion.label ? suggestion.label : '');

const Autosuggest: React.FC<Props<Suggestion>> = props => {
  const inputId = useUuid(props.inputProps.id);
  const errorTextUuid = useUuid(props.errorTextId);

  const inputProps: InputProps<Suggestion> = {
    ...props.inputProps,
    id: inputId,
    'aria-invalid': props.error,
    'aria-describedby': getAriaDescribedBy({ ...props.inputProps, ...props }, errorTextUuid),
  };

  return (
    <ErrorWrapper errorText={props.errorText} errorTextId={errorTextUuid} className={props.errorWrapperClassName}>
      <div className={props.className}>
        {renderLabel(props.label, inputId, props.error ? FormMode.oninvalid : FormMode.onwhite)}
        <ReactAutosuggest
          theme={styles}
          suggestions={props.suggestions}
          inputProps={inputProps}
          onSuggestionsFetchRequested={props.onSuggestionsFetchRequested}
          onSuggestionsClearRequested={props.onSuggestionsClearRequested}
          onSuggestionSelected={props.onSuggestionSelected}
          getSuggestionValue={getSuggestionValue}
          renderSuggestion={props.renderSuggestion ? props.renderSuggestion : renderSuggestion}
          focusInputOnSuggestionClick={props.focusInputOnSuggestionClick || false}
        />
      </div>
    </ErrorWrapper>
  );
};

export default Autosuggest;
