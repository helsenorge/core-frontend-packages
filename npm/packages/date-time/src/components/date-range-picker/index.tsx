import * as React from 'react';

import moment from 'moment';
import 'react-dates/initialize';
import { FocusedInputShape, SingleDatePickerShape, DateRangePickerShape } from 'react-dates';

import { LanguageLocales } from '@helsenorge/core-utils/constants/languages';
import { isMobileUA } from '@helsenorge/core-utils/user-agents-utils';

import ValidationError from '@helsenorge/form/components/molecules/form/validation-error';
import {
  DEFAULT_DATE_PLACEHOLDER_NB,
  DEFAULT_DATE_PLACEHOLDER_EN,
  DEFAULT_STARTDATE_PLACEHOLDER,
  DEFAULT_ENDDATE_PLACEHOLDER,
  DEFAULT_DATE_PLACEHOLDER_SE,
} from './../../../constants/datetime';
import { renderDesktopDatePicker } from './date-range-picker-desktop';
import { DateRangePickerLabel } from './date-range-picker-label';
import { renderMobileDatePicker } from './date-range-picker-mobile';
import { DateRangePickerDefaultProps, DateRangePickerProps, DateRangePickerState } from './date-range-picker-types';
import {
  START_DATE_ID,
  END_DATE_ID,
  Phrases,
  ErrorPhrases,
  getCSSClasses,
  getDateNOString,
  notifyValidated,
  getDefaultVisibleMonth,
} from './date-range-picker-utils';
import { validateSingleDate, validateRangeDate, isPrevMonthValid, isNextMonthValid } from './date-range-picker-validation';

import 'react-dates/lib/css/_datepicker.css';
import toolkitstyles from './styles.module.scss';

export class DateRangePicker extends React.Component<DateRangePickerProps, DateRangePickerState> {
  public static defaultProps: DateRangePickerDefaultProps = {
    id: 'daterangepicker',
    resources: Phrases,
    errorResources: ErrorPhrases,
    startDateId: START_DATE_ID,
    startDatePlaceholderText: DEFAULT_STARTDATE_PLACEHOLDER,
    endDateId: END_DATE_ID,
    endDatePlaceholderText: DEFAULT_ENDDATE_PLACEHOLDER,
    showDefaultInputIcon: true,
    isValidationHidden: false,
    isRequired: false,
    isDisabled: false,
    numberOfMonths: 1,
  };

  airbnbSingleDatepickerRef: React.RefObject<React.ClassicComponent<SingleDatePickerShape, {}>>;
  airbnbDateRangepickerRef: React.RefObject<React.ClassicComponent<DateRangePickerShape, {}>>;

  constructor(props: DateRangePickerProps) {
    super(props);

    const momentLocale = moment();
    if (props.locale === LanguageLocales.ENGLISH) {
      require('moment/locale/en-gb');
      momentLocale.locale('en-gb');
    } else if (props.locale === LanguageLocales.SAMI_NORTHERN) {
      require('moment/locale/se');
      momentLocale.locale('se');
    } else {
      require('moment/locale/nb');
      momentLocale.locale('nb');
    }

    this.airbnbSingleDatepickerRef = React.createRef();
    this.airbnbDateRangepickerRef = React.createRef();

    let placeholder = '';

    switch (props.locale) {
      case LanguageLocales.ENGLISH:
        placeholder = DEFAULT_DATE_PLACEHOLDER_EN;
        break;
      case LanguageLocales.SAMI_NORTHERN:
        placeholder = DEFAULT_DATE_PLACEHOLDER_SE;
        break;
      default:
        placeholder = DEFAULT_DATE_PLACEHOLDER_NB;
    }

    this.state = {
      isMobile: isMobileUA(),
      momentLocale: momentLocale.localeData(),
      placeholder: props.placeholder ? props.placeholder : placeholder,
      singleDate: props.singleDateValue ?? null,
      isSingleDateFocused: null,
      isSingleDateValid: true,
      startDate: props.startDateValue ?? null,
      endDate: props.endDateValue ?? null,
      isRangeDateValid: true,
      validated: false,
      focusedInput: null,
      visibleMonth: getDefaultVisibleMonth(props),
    };
  }

  /** Håndterer deaktivering av måned knappene hvis de minimum/maximumDate er satt */
  componentDidMount() {
    this.disableMonthSelectorHandler(this.state.visibleMonth);
  }

  /** Hvis valuen som kommer fra prop er endret underveis så må den gå gjennom validator */
  componentDidUpdate(prevProps: DateRangePickerProps, prevState: DateRangePickerState): void {
    if (prevProps.type === 'single') {
      const prevPropSingleDateValue = getDateNOString(prevProps.singleDateValue);
      const prevStateSingleDate = getDateNOString(prevState.singleDate);
      const currentSingleDateValue = getDateNOString(this.props.singleDateValue);
      if (prevPropSingleDateValue !== currentSingleDateValue && prevStateSingleDate !== currentSingleDateValue) {
        this.setState(
          {
            singleDate: this.props.singleDateValue ?? null,
            visibleMonth: getDefaultVisibleMonth(this.props),
          },
          () => {
            if (!this.state.isSingleDateFocused) {
              this.singleDateValidator(this.state.singleDate);
            }
          }
        );
      }
    } else {
      const prevPropStartDateValue = getDateNOString(prevProps.startDateValue);
      const prevStateStartDate = getDateNOString(prevState.startDate);
      const currentStartDateValue = getDateNOString(this.props.startDateValue);
      const prevPropEndDateValue = getDateNOString(prevProps.endDateValue);
      const prevStateEndDate = getDateNOString(prevState.endDate);
      const currentEndDateValue = getDateNOString(this.props.endDateValue);
      if (prevPropStartDateValue !== currentStartDateValue && prevStateStartDate !== currentStartDateValue) {
        this.setState(
          {
            startDate: this.props.startDateValue ?? null,
          },
          () => {
            if (this.state.focusedInput === null) {
              this.rangeDateValidator(this.state.startDate, this.state.endDate);
            }
          }
        );
      }
      if (prevPropEndDateValue !== currentEndDateValue && prevStateEndDate !== currentEndDateValue) {
        this.setState({ endDate: this.props.endDateValue ?? null }, () => {
          if (this.state.focusedInput === null) {
            this.rangeDateValidator(this.state.startDate, this.state.endDate);
          }
        });
      }
    }

    /* Sjekker om nåværende måned vi ser har endret seg, eller om det skal være lov å gå lenger forover/bakover i tid */
    if (
      prevState.visibleMonth !== this.state.visibleMonth ||
      (prevProps.maximumDate && this.props.maximumDate && !prevProps.maximumDate.isSame(this.props.maximumDate, 'month')) ||
      (prevProps.minimumDate && this.props.minimumDate && !prevProps.minimumDate.isSame(this.props.minimumDate, 'month'))
    ) {
      this.disableMonthSelectorHandler(this.state.visibleMonth);
    }
  }

  /* Calling the validation on the incoming date */
  onSingleDateChange = (singleDate: moment.Moment): void => {
    this.setState({ singleDate });
    this.props.onDateChange && this.props.onDateChange(singleDate, this.props.id);
  };

  /* Calling the validation on the incoming date with the stored end date */
  onStartDateChange = (startDate: moment.Moment): void => {
    this.setState({ startDate });
    this.props.onDateChange && this.props.onDateChange({ start: startDate, end: this.state.endDate }, this.props.id);
  };

  /* Calling the validation on the incoming date with the stored start date */
  onEndDateChange = (endDate: moment.Moment): void => {
    this.setState({ endDate });
    this.props.onDateChange && this.props.onDateChange({ start: this.state.startDate, end: endDate }, this.props.id);
  };

  onRangeDatesChange = ({ startDate, endDate }: { startDate: moment.Moment; endDate: moment.Moment }): void => {
    this.setState({ startDate, endDate });
    this.props.onDateChange && this.props.onDateChange({ start: startDate, end: endDate }, this.props.id);
  };

  singleDateValidator = (date: moment.Moment | null): Promise<void> => {
    const { id, isRequired, errorResources, minimumDate, maximumDate, dateValidator, onValidated, onError } = this.props;

    const v: {
      isSingleDateValid: boolean;
      errorString?: string;
    } = validateSingleDate(this.airbnbSingleDatepickerRef, date, id, isRequired, errorResources, minimumDate, maximumDate, dateValidator);

    return new Promise<void>((resolve: () => void) => {
      this.setState(
        {
          validated: true,
          isSingleDateValid: v.isSingleDateValid,
          errorString: v.errorString,
        },
        () => {
          notifyValidated(id, date, v.isSingleDateValid, v.errorString, onValidated, onError);
          resolve();
        }
      );
    });
  };

  rangeDateValidator = (startDate: moment.Moment | null, endDate: moment.Moment | null): Promise<void> => {
    const { id, isRequired, errorResources, minimumDate, maximumDate, minimumPeriod, dateValidator, onValidated, onError } = this.props;

    const v: {
      isRangeDateValid: boolean;
      errorString?: string;
    } = validateRangeDate(
      this.airbnbDateRangepickerRef,
      startDate,
      endDate,
      id,
      isRequired,
      errorResources,
      minimumDate,
      maximumDate,
      minimumPeriod,
      dateValidator
    );

    return new Promise<void>((resolve: () => void) => {
      this.setState(
        {
          validated: true,
          isRangeDateValid: v.isRangeDateValid,
          errorString: v.errorString,
        },
        () => {
          notifyValidated(id, { start: startDate, end: endDate }, v.isRangeDateValid, v.errorString, onValidated, onError);
          resolve();
        }
      );
    });
  };

  onSingleDateFocusChange = ({ focused }: { focused: boolean | null }): void => {
    // This replaces onBlur method and validates the field for everytime the focus is brought outside
    // setTimeout is needed to delay the check until the focus is moved
    if (!focused) {
      setTimeout(() => {
        this.validateField();
      }, 0);
    }
    if (focused !== this.state.isSingleDateFocused) {
      this.setState({ isSingleDateFocused: focused });
    }
  };

  onRangeDatesFocusChange = (focusedInput: FocusedInputShape): void => {
    // This replaces onBlur method and validates the field for everytime the focus is brought outside
    // setTimeout is needed to delay the check until the focus is moved
    setTimeout(() => {
      if (this.state.focusedInput === null) {
        this.validateField();
        this.setState({ isFocusReset: true });
      }
    }, 0);

    this.setState({ focusedInput });

    // Rekalkulerer om navNext og navPrev skal være disabled
    if (this.state.isFocusReset) {
      this.setState({ isFocusReset: false });
    }
  };

  onCalenderIconClick = (): void => {
    if (this.state.isSingleDateFocused) {
      this.setState({ isSingleDateFocused: false });
    } else if (this.state.focusedInput) {
      this.setState({ focusedInput: null });
    }

    this.setState({ isFocusReset: true });
  };

  /* Runs dates validator */
  validateField(): Promise<void> {
    return this.props.type === 'single'
      ? this.singleDateValidator(this.state.singleDate)
      : this.rangeDateValidator(this.state.startDate, this.state.endDate);
  }

  /* Returns whereas the field is valid (in a form context) */
  isValid(): boolean {
    return this.props.type === 'single' ? this.state.isSingleDateValid : this.state.isRangeDateValid;
  }

  /* returns the error string, used by date-time-picker to access date errors */
  getErrorString = (): string | undefined => this.state.errorString;

  /* onBlurHandler for NativeInput only (desktop uses airbnb focus methods) */
  onBlurHandler = (): void => {
    this.validateField();
    this.setState({ isFocusReset: true });
  };

  /* Disables/enables next and last month selector buttons based on maximum/minimumDate */
  disableMonthSelectorHandler = (date: moment.Moment): void => {
    const disabledNext = !isNextMonthValid(date, this.props.maximumDate);
    this.setState({ isNextMonthDisabled: disabledNext });
    const disabledPrev = !isPrevMonthValid(date, this.props.minimumDate);
    this.setState({ isPrevMonthDisabled: disabledPrev });
  };

  onChangeVisibleMonthHandler = (date: moment.Moment) => {
    this.setState({ visibleMonth: date });
  };

  render(): JSX.Element {
    const {
      locale,
      className,
      label,
      isLabelHidden,
      requiredLabel,
      optionalLabel,
      subLabel,
      helpButton,
      helpElement,
      isValidationHidden,
      validationErrorRenderer,
      ...partialPropsForDesktop
    } = this.props;

    const { isSingleDateValid, isRangeDateValid, errorString } = this.state;
    const { id, type, isRequired } = partialPropsForDesktop;
    const hasErrors = type === 'single' ? !isSingleDateValid : type === 'range' ? !isRangeDateValid : false;
    const classes = getCSSClasses(
      toolkitstyles.datepicker,
      toolkitstyles['datepicker--haserror'],
      !isValidationHidden,
      hasErrors,
      className
    );
    moment.locale(locale);

    return (
      <div id={`${id}-wrapper`} className={classes}>
        <fieldset>
          {hasErrors && validationErrorRenderer}
          {!isValidationHidden && hasErrors && <ValidationError isValid={!hasErrors} error={errorString ? errorString : 'Error'} />}
          {label && (
            <DateRangePickerLabel
              label={label}
              locale={locale}
              isLabelHidden={isLabelHidden}
              isRequired={isRequired}
              requiredLabel={requiredLabel}
              optionalLabel={optionalLabel}
              helpButton={helpButton}
              subLabel={subLabel}
            />
          )}
          {helpElement ?? null}
          {this.state.isMobile
            ? renderMobileDatePicker(
                this.props,
                this.state,
                this.onSingleDateChange,
                this.onStartDateChange,
                this.onEndDateChange,
                this.onBlurHandler
              )
            : renderDesktopDatePicker(
                partialPropsForDesktop,
                this.state,
                this.airbnbSingleDatepickerRef,
                this.airbnbDateRangepickerRef,
                this.onSingleDateChange,
                this.onSingleDateFocusChange,
                this.onRangeDatesChange,
                this.onRangeDatesFocusChange,
                this.onCalenderIconClick,
                this.onChangeVisibleMonthHandler
              )}
        </fieldset>
      </div>
    );
  }
}

export default DateRangePicker;
