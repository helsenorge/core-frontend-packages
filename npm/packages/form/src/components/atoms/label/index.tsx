import * as React from 'react';

import classNames from 'classnames';

import { Sublabel } from './sublabel';

import toolkitstyles from './styles.module.scss';

interface Props {
  /** Tekst eller Element som vises i <label> node */
  labelText: string | JSX.Element;
  /** htmlFor HTML property som settes på <label> node */
  htmlFor?: string;
  /** Tekst eller Element som vises i <label> node */
  sublabelText?: string | JSX.Element;
  /** Element for hjelpeknapp */
  helpButton?: JSX.Element;
  /* CSS-classes som legges på selve knappen */
  className?: string;
  /** Legger CSSclasse for å styre font-weight på label  */
  isNotBold?: boolean;
  /** Optional ref som forwardes til HTMLButtonElement */
  ref?: React.RefObject<HTMLDivElement>;
  /** Id som benyttes for å hente ut komponent i automatiske tester */
  testId?: string;
}

export const Label = React.forwardRef<HTMLDivElement, React.PropsWithChildren<Props>>((
  { labelText, isNotBold, htmlFor, sublabelText, helpButton, className, children, testId },
  ref: React.RefObject<HTMLDivElement>
) => {
  const labelRef = React.useRef<HTMLDivElement>(null);

  const labelClasses = classNames(toolkitstyles['label'], { [toolkitstyles['label--isnotbold']]: isNotBold });
  return (
    <div ref={ref ? ref : labelRef} className={className ? className : ''}>
      <label className={labelClasses} htmlFor={htmlFor} data-testid={testId}>
        {labelText}
        {helpButton}
        {sublabelText && <Sublabel sublabelText={sublabelText} />}
      </label>
      {children}
    </div>
  );
});

export default Label;
