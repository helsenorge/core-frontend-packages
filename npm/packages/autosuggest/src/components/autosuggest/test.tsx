import * as React from 'react';

import { render, screen } from '@testing-library/react';
import { mount } from 'enzyme';

import { Label } from '@helsenorge/form/components/label';

import Autosuggest, { Suggestion } from '.';

describe('Autosuggest', () => {
  const suggestions: Array<Suggestion> = [
    { label: 'First suggestion', value: 'First suggestion' },
    { label: 'Second suggestion', value: 'Second suggestion', optionalLabel: 'optionalLabel' },
    { label: 'Third suggestion', value: 'Third suggestion' },
    { label: 'Fourth suggestion', value: 'Fourth suggestion' },
  ];

  describe('Gitt at Autosuggest er initialisert', () => {
    const onChangeMock = jest.fn();
    const onBlurMock = jest.fn();
    const onSuggestionsFetchRequestedMock = jest.fn();
    const onSuggestionsClearRequestedMock = jest.fn();
    const onSuggestionsSelectedMock = jest.fn();
    const onSubmitValidatorMock = jest.fn();

    describe('Når Autosuggest vises', () => {
      it('Så finnes det et søkefelt med label', () => {
        render(
          <Autosuggest
            id={'id1'}
            type="search"
            label={'Søk'}
            value={'default value'}
            suggestions={[]}
            onChange={onChangeMock}
            onSuggestionsFetchRequested={onSuggestionsFetchRequestedMock}
            onSuggestionsClearRequested={onSuggestionsClearRequestedMock}
            onSuggestionSelected={onSuggestionsSelectedMock}
            noValidation
          />
        );

        const input = screen.getByLabelText('Søk');
        expect(input).toBeVisible();
      });
    });

    describe('Når den mountes med labels og noValidation', () => {
      const wrapper = mount(
        <Autosuggest
          id={'id1'}
          type="search"
          label={'This is an autosuggest (input search) field with label'}
          subLabel={'and its sublabel'}
          value={'default value'}
          suggestions={[]}
          onChange={onChangeMock}
          onSuggestionsFetchRequested={onSuggestionsFetchRequestedMock}
          onSuggestionsClearRequested={onSuggestionsClearRequestedMock}
          onSuggestionSelected={onSuggestionsSelectedMock}
          noValidation
        />
      );

      it('Så rendres det Label og sublabel + autosuggest component', () => {
        expect(wrapper.find(Label).length).toEqual(1);
        expect(wrapper.render()).toMatchSnapshot();
      });

      it('Så kalles det ikke validate ved onChange', () => {
        const instance: Autosuggest = wrapper.instance() as Autosuggest;
        jest.spyOn(instance, 'validate');

        wrapper.find('input').simulate('change', { target: { name: 'name_of_input', value: 'value' } });
        wrapper.update();
        expect(onChangeMock).toHaveBeenCalled();
        expect(instance.validate).not.toHaveBeenCalled();
      });
    });

    describe('Når den mountes med validation', () => {
      const wrapper = mount(
        <Autosuggest
          id={'id1'}
          type="search"
          label={'This is an autosuggest (input search) field with label'}
          subLabel={'and its sublabel'}
          value={'default value'}
          suggestions={[]}
          onChange={onChangeMock}
          onSuggestionsFetchRequested={onSuggestionsFetchRequestedMock}
          onSuggestionsClearRequested={onSuggestionsClearRequestedMock}
          onSuggestionSelected={onSuggestionsSelectedMock}
          onSubmitValidator={onSubmitValidatorMock}
          maxLength={10}
          isRequired
          noValidation
        />
      );
      const instance: Autosuggest = wrapper.instance() as Autosuggest;

      it('Så isValid er oppdatert iht staten', () => {
        expect(instance.isValid()).toEqual(true);
      });

      it('Så validate oppdaterer staten iht sjekkene', () => {
        // does not validate empty string when required
        expect(instance.validate('')).toEqual(false);
        // does not validate invalid characters
        expect(instance.validate('#')).toEqual(false);
        // does validate long string
        expect(instance.validate('ølkjhgfdsdfghjklølkjhgfdsdfghjkllkjhgfdsdfghjklkjhgfdfgh')).toEqual(false);
      });

      it('Så kalles det onSubmitValidatorMock når angitt', () => {
        instance.validateField();
        expect(onSubmitValidatorMock).toHaveBeenCalled();
      });
    });

    describe('Når det gjøres endringer i input feltet', () => {
      const wrapper = mount(
        <Autosuggest
          id={'id2'}
          type="search"
          value={'default value'}
          suggestions={[]}
          onChange={onChangeMock}
          onSuggestionsFetchRequested={onSuggestionsFetchRequestedMock}
          onSuggestionsClearRequested={onSuggestionsClearRequestedMock}
          onSuggestionSelected={onSuggestionsSelectedMock}
          errorMessage={'error message'}
          isRequired
          minLength={2}
          maxLength={50}
        />
      );
      const instance: Autosuggest = wrapper.instance() as Autosuggest;

      it('Så kalles det onChange og validate', () => {
        jest.spyOn(instance, 'validate');

        wrapper.find('input').simulate('change', { target: { name: 'name_of_input', value: 'value' } });
        wrapper.update();
        expect(onChangeMock).toHaveBeenCalled();
        expect(instance.validate).toHaveBeenCalled();
      });
    });

    describe('Når det settes focus på inputFeltet', () => {
      const wrapper = mount(
        <Autosuggest
          id={'id3'}
          type="search"
          value={'default value'}
          suggestions={[]}
          onBlur={onBlurMock}
          onChange={onChangeMock}
          onSuggestionsFetchRequested={onSuggestionsFetchRequestedMock}
          onSuggestionsClearRequested={onSuggestionsClearRequestedMock}
          onSuggestionSelected={onSuggestionsSelectedMock}
          errorMessage={'error message'}
          isRequired
          minLength={2}
          maxLength={50}
        />
      );
      const instance: Autosuggest = wrapper.instance() as Autosuggest;

      it('Så kalles det onBlur class method i tillegg til prop', () => {
        jest.useFakeTimers();
        jest.spyOn(instance, 'onBlur');

        wrapper.find('input').simulate('change', { target: { value: '1' } });
        wrapper.find('input').simulate('blur');
        jest.runAllTimers();

        wrapper.update();
        expect(instance.onBlur).toHaveBeenCalled();
        expect(onBlurMock).toHaveBeenCalled();
      });
    });

    describe('Når det brukes tastaturet', () => {
      const wrapper = mount(
        <Autosuggest
          id={'id3'}
          type="search"
          value={'default value'}
          suggestions={[]}
          onChange={onChangeMock}
          onSuggestionsFetchRequested={onSuggestionsFetchRequestedMock}
          onSuggestionsClearRequested={onSuggestionsClearRequestedMock}
          onSuggestionSelected={onSuggestionsSelectedMock}
          errorMessage={'error message'}
          isRequired
          minLength={2}
          maxLength={50}
        />
      );
      const instance: Autosuggest = wrapper.instance() as Autosuggest;

      it('Så håndteres det Keydown riktig', () => {
        jest.useFakeTimers();
        jest.spyOn(instance, 'onKeyDown');
        wrapper.find('input').simulate('change', { target: { value: 'abcdefg' } });
        wrapper.find('input').simulate('keydown', { keyCode: 13 });

        jest.runAllTimers();

        wrapper.update();
        expect(instance.onKeyDown).toHaveBeenCalled();
      });
    });

    describe('Når det rendres suggestions', () => {
      const wrapper = mount(
        <Autosuggest
          id={'id4'}
          type="search"
          value={'default value'}
          suggestions={suggestions}
          onBlur={onBlurMock}
          onChange={onChangeMock}
          onSuggestionsFetchRequested={onSuggestionsFetchRequestedMock}
          onSuggestionsClearRequested={onSuggestionsClearRequestedMock}
          onSuggestionSelected={onSuggestionsSelectedMock}
          errorMessage={'error message'}
          isRequired
        />
      );

      it('Så rendres det riktig markup med og uten optionalLabel', () => {
        const instance: Autosuggest = wrapper.instance() as Autosuggest;
        const suggestionMarkup = instance.renderSuggestion(suggestions[0]);
        const suggestionMarkupWithOptionalLabel = instance.renderSuggestion(suggestions[1]);
        expect(suggestionMarkup).toMatchSnapshot();
        expect(suggestionMarkupWithOptionalLabel).toMatchSnapshot();
      });
    });

    describe('Når den mountes med noCharacterValidation', () => {
      const wrapper = mount(
        <Autosuggest
          id={'id5'}
          type="search"
          label={'This is an autosuggest (input search) field with label'}
          subLabel={'and its sublabel'}
          value={'default value'}
          suggestions={[]}
          onChange={onChangeMock}
          onSuggestionsFetchRequested={onSuggestionsFetchRequestedMock}
          onSuggestionsClearRequested={onSuggestionsClearRequestedMock}
          onSuggestionSelected={onSuggestionsSelectedMock}
          onSubmitValidator={onSubmitValidatorMock}
          maxLength={10}
          isRequired
          noCharacterValidation={true}
        />
      );
      const instance: Autosuggest = wrapper.instance() as Autosuggest;

      it('Så validate oppdaterer staten iht sjekkene', () => {
        // does not validate empty string when required
        expect(instance.validate('')).toEqual(false);
        // does not validate invalid characters
        expect(instance.validate('#')).toEqual(true);
        // does validate long string
        expect(instance.validate('ølkjhgfdsdfghjklølkjhgfdsdfghjkllkjhgfdsdfghjklkjhgfdfgh')).toEqual(true);
      });

      it('Så kalles det onSubmitValidatorMock når angitt', () => {
        instance.validateField();
        expect(onSubmitValidatorMock).toHaveBeenCalled();
      });
    });
  });
});
