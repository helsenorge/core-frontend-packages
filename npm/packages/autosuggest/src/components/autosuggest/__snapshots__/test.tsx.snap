// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autosuggest Gitt at Autosuggest er initialisert Når den mountes med labels og noValidation Så rendres det Label og sublabel + autosuggest component 1`] = `
<div
  class=""
  id="id1-wrapper"
>
  <div
    class=""
  >
    <label
      class="label"
      for="id1"
    >
      This is an autosuggest (input search) field with label
      <span
        class="label__sublabel"
      >
        and its sublabel
      </span>
    </label>
  </div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-owns="react-autowhatever-1"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="react-autowhatever-1"
      autocomplete="off"
      id="id1"
      placeholder="Placeholder text"
      type="search"
      value="default value"
    />
    <div
      id="react-autowhatever-1"
      role="listbox"
    />
  </div>
</div>
`;

exports[`Autosuggest Gitt at Autosuggest er initialisert Når det rendres suggestions Så rendres det riktig markup med og uten optionalLabel 1`] = `
<span>
  First suggestion
</span>
`;

exports[`Autosuggest Gitt at Autosuggest er initialisert Når det rendres suggestions Så rendres det riktig markup med og uten optionalLabel 2`] = `
<span>
  Second suggestion
  <em>
     
    optionalLabel
  </em>
</span>
`;
